### DISCUSSION TITLE
Make custom component behave differently between training and test

### DISCUSSION THREAD

**Discussion by exp-ouroborous** [explosion/spaCy#8148] [Help: Coding & Implementations] [answered]
Hello Folks,
I have a pipeline where I use a custom PhraseMatcher and the builtin Spacy NER model to identify NEs. I want to turn "off" the custom  PhraseMatcher component during training as I want the built in NER model to get trained on everything. What is the recommended way to achieve this?

[This page](https://spacy.io/usage/training#config-components) mentions frozen components but that wont work because I actually want the component to be completely "off".

Ideally if there was something I could check in the custom component code as to whether we are in training or not - I could use that to turn the component on/off. I did some spelunking in the source code and found a few mentions of an `is_train` parameter in the `forward` call for Thinc models. But I could not figure out how to use something similar in the custom component.

I understand that I could do this programmatically in python but given v3's suggestion to use cli + config files I was trying to see if this doable using just that.

**Comment by polm** [comment, marked as answer]
You can just list a component as "disabled" in the nlp block.

```
[nlp]
... other stuff...
disabled = ["entity_ruler"]
```

A frozen component won't be trained but will run, while a frozen component will be in the pipeline but won't do anything. At inference time, after loading the config, you can re-enable the component using the `nlp.enable_pipe` function. See [here](https://spacy.io/api/data-formats#:~:text=names%20of%20pipeline%20components%20that%20are%20loaded%20but%20disabled%20by%20default%20and%20not%20run%20as%20part%20of%20the%20pipeline.%20) for docs on that.

**Reply by exp-ouroborous** [reply]
That works but does need me to change the pipeline in python during inference (admittedly not much code). I was wondering if there is a way to do this such that I can use the cli `spacy evaluate` directly when testing my model.

Another reason to be able to do this within the component is if I wanted to simulate a "dropout". So I wanted to experiment with turning my custom PhraseMatcher on/off randomly based on a dropout rate. So this dropout would happen only during training but the component ought to always be on during testing.

**Reply by polm** [reply]
Since you can override the config from the command line, you could pass `spacy evaluate` an option like `--nlp.disabled="[]"` to re-enable the component as needed.

If you want to do something like your dropout example, that's necessarily more complicated. You're right that Thinc models use `is_train`, but the pipeline component that calls them doesn't pass that value directly - instead there are methods that are used in training (like `update`) that call training-specific model methods (like `begin_update`). I don't think we have an example or guide for something like that, so you'll have to do some source diving, but the dropout implementation in Thinc is pretty short, so it should be easy to use as a reference.

**Reply by exp-ouroborous** [reply]
The cli argument is a great suggestion. Thank you!
